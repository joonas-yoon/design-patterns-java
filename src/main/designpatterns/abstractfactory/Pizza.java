package main.designpatterns.abstractfactory;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

public abstract class Pizza {

    protected String name;
    protected String dough;
    protected String sauce;
    protected String cheese;
    protected String veggies;
    protected List<String> toppings = new ArrayList<>();

    public abstract void prepare();

    public void setName(String name) {
        this.name = name;
    }

    public void cook() {
        if (sauce != null)
            sauce += " (grilled)";
        System.out.println("Cook: " + this);
    }

    public void cut() {
        if (dough != null)
            dough += " (sliced)";
        System.out.println("Cut: " + this);
    }

    @Override
    public String toString() {
        List<String> ingredients = List.of(
                Optional.ofNullable(dough).orElse("No dough"),
                Optional.ofNullable(sauce).orElse("No sauce"),
                Optional.ofNullable(cheese).orElse("No cheese"),
                Optional.ofNullable(veggies).orElse("No veggies")
        );
        return String.format("Pizza {name=%s, ingredients=%s, toppings=%s}",
                name, ingredients, Arrays.toString(toppings.toArray()));
    }
}
