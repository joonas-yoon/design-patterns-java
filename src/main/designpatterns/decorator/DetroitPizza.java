package main.designpatterns.decorator;

import java.util.Arrays;

public class DetroitPizza extends Pizza {

    public DetroitPizza() {
        toppings = Arrays.asList(
                "Tomato sauce",
                "Pepperoni",
                "Mozzarella cheese",
                "Cheddar cheese");
        dough.setShape(Dough.Shape.RECTANGLE);
    }

    @Override
    public int getPrice() {
        return 6000;
    }
}
