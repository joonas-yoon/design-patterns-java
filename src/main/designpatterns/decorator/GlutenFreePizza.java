package main.designpatterns.decorator;

import java.util.List;

public class GlutenFreePizza extends PizzaTopper {

    private static final int GLUTEN_FREE_PRICE = 4000;

    public GlutenFreePizza(Pizza pizza) {
        this.pizza = pizza;
    }

    @Override
    public int getPrice() {
        return pizza.getPrice() + GLUTEN_FREE_PRICE;
    }

    @Override
    public Dough getDough() {
        Dough newDough = pizza.getDough();
        newDough.setStyle(Dough.Style.GLUTEN_FREE);
        return newDough;
    }

    @Override
    public List<String> getToppings() {
        return pizza.getToppings();
    }
}
