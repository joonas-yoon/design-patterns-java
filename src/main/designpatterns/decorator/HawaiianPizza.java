package main.designpatterns.decorator;

import java.util.Arrays;

public class HawaiianPizza extends Pizza {

    public HawaiianPizza() {
        toppings = Arrays.asList("Tomato sauce", "Pineapple", "Bacon", "Cheese");
    }

    @Override
    public int getPrice() {
        return 8000;
    }
}
