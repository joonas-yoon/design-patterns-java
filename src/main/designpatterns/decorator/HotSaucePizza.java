package main.designpatterns.decorator;

import java.util.List;

public class HotSaucePizza extends PizzaTopper {

    private static final int HOT_SAUCE_PRICE = 300;

    public HotSaucePizza(Pizza pizza) {
        this.pizza = pizza;
    }

    @Override
    public int getPrice() {
        return pizza.getPrice() + HOT_SAUCE_PRICE;
    }

    @Override
    public List<String> getToppings() {
        List<String> toppings = pizza.getToppings();
        toppings.add("Hot sauce");
        return toppings;
    }
}
