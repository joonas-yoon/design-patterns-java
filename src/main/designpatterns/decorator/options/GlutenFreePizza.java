package main.designpatterns.decorator.options;

import java.util.List;
import main.designpatterns.decorator.Dough;
import main.designpatterns.decorator.Pizza;
import main.designpatterns.decorator.PizzaTopper;

public class GlutenFreePizza extends PizzaTopper {

    private static final int GLUTEN_FREE_PRICE = 4000;

    public GlutenFreePizza(Pizza pizza) {
        this.pizza = pizza;
    }

    @Override
    public int getPrice() {
        return pizza.getPrice() + GLUTEN_FREE_PRICE;
    }

    @Override
    public Dough getDough() {
        Dough newDough = pizza.getDough();
        newDough.setStyle(Dough.Style.GLUTEN_FREE);
        return newDough;
    }

    @Override
    public List<String> getToppings() {
        return pizza.getToppings();
    }
}
