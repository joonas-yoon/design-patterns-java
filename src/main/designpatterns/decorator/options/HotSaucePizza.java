package main.designpatterns.decorator.options;

import java.util.List;
import main.designpatterns.decorator.Pizza;
import main.designpatterns.decorator.PizzaTopper;

public class HotSaucePizza extends PizzaTopper {

    private static final int HOT_SAUCE_PRICE = 300;

    public HotSaucePizza(Pizza pizza) {
        this.pizza = pizza;
    }

    @Override
    public int getPrice() {
        return pizza.getPrice() + HOT_SAUCE_PRICE;
    }

    @Override
    public List<String> getToppings() {
        List<String> toppings = pizza.getToppings();
        toppings.add("Hot sauce");
        return toppings;
    }
}
