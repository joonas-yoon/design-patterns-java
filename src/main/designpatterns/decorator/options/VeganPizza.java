package main.designpatterns.decorator.options;

import java.util.List;
import java.util.Set;
import main.designpatterns.decorator.Pizza;
import main.designpatterns.decorator.PizzaTopper;

public class VeganPizza extends PizzaTopper {

    private static final int VEGAN_PRICE = 3000;

    public VeganPizza(Pizza pizza) {
        this.pizza = pizza;
    }

    @Override
    public int getPrice() {
        return pizza.getPrice() + VEGAN_PRICE;
    }

    @Override
    public List<String> getToppings() {
        List<String> toppings = pizza.getToppings();
        toppings.removeAll(Set.of("Ham", "Bacon"));
        toppings.replaceAll(topping -> {
            boolean hasCheese = topping.toLowerCase().contains("cheese");
            if (!hasCheese) {
                return topping;
            }
            return "Vegan cheese";
        });
        return toppings;
    }
}