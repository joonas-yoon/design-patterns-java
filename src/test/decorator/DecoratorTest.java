package test.decorator;

import main.designpatterns.decorator.DetroitPizza;
import main.designpatterns.decorator.GlutenFreePizza;
import main.designpatterns.decorator.HawaiianPizza;
import main.designpatterns.decorator.HotSaucePizza;
import main.designpatterns.decorator.Pizza;
import main.designpatterns.decorator.VeganPizza;
import org.junit.jupiter.api.Test;
import test.BaseUnitTest;

class DecoratorTest extends BaseUnitTest {

    @Test
    public void testDecorateTopping() {
        println("# Create new HawaiianPizza");
        Pizza pizza = new HawaiianPizza();
        println(toString(pizza));

        println("Add HotSauced option");
        HotSaucePizza hotSaucedPizza = new HotSaucePizza(pizza);
        println(toString(hotSaucedPizza));

        println("Add HotSauced option");
        HotSaucePizza superHotPizza = new HotSaucePizza(hotSaucedPizza);
        println(toString(superHotPizza));

        println("Add ForVegan option");
        VeganPizza superHotVeganPizza = new VeganPizza(superHotPizza);
        println(toString(superHotVeganPizza));
    }

    @Test
    public void testDecorateDough() {
        println("# Create new DetroitPizza");
        Pizza pizza2 = new DetroitPizza();
        println(toString(pizza2));

        println("Add GlutenFree option");
        GlutenFreePizza glutenFreePizza2 = new GlutenFreePizza(pizza2);
        println(toString(glutenFreePizza2));
    }

    private static String toString(Pizza pizza) {
        return pizza + ", price=" + pizza.getPrice() + "\n";
    }
}