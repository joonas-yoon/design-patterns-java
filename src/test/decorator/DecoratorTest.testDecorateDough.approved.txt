# Create new DetroitPizza
{Dough=PLAIN/RECTANGLE, Toppings=[Tomato sauce, Pepperoni, Mozzarella cheese, Cheddar cheese]}, price=6000

Add GlutenFree option
{Dough=GLUTEN_FREE/RECTANGLE, Toppings=[Tomato sauce, Pepperoni, Mozzarella cheese, Cheddar cheese]}, price=10000

